middleware/cluster.go

```diff
@@ -0,0 +1,45 @@
+package middleware
+
+import (
+	"fmt"
+	"net/http"
+	"regexp"
+	"strings"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/grafana/dskit/clusterutil"
+)
+
+// ClusterValidationMiddleware validates that requests have the correct cluster verification label.
+func ClusterValidationMiddleware(cluster string, auxPaths []string, invalidClusters *prometheus.CounterVec, logger log.Logger) Interface {
+	var reB strings.Builder
+	// Allow for a potential path prefix being configured.
+	reB.WriteString(".*/(metrics|debug/pprof.*|ready")
+	for _, p := range auxPaths {
+		reB.WriteString("|" + regexp.QuoteMeta(p))
+	}
+	reB.WriteString(")")
+	reAuxPath := regexp.MustCompile(reB.String())
+
+	return Func(func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			reqCluster := r.Header.Get(clusterutil.ClusterVerificationLabelHeader)
+			if !reAuxPath.MatchString(r.URL.Path) && reqCluster != cluster {
+				level.Warn(logger).Log("msg", "rejecting request with wrong cluster verification label",
+					"cluster_verification_label", cluster, "request_cluster_verification_label", reqCluster,
+					"header", clusterutil.ClusterVerificationLabelHeader, "url", r.URL, "path", r.URL.Path)
+				if invalidClusters != nil {
+					invalidClusters.WithLabelValues("http", r.URL.Path, reqCluster).Inc()
```

Comment:
I think this is very fragile, the's no contract on which labels `invalidClusters` has, and if those change, this will panic when it should fail the request. I would suggest passing here a closure that would increment the correct metric (and let the caller pass an empty closure, not nil), so the responsibility of declaring labels and using them would stay on the initialization side.

Shameless plug, this is why I wrote [cabify/gotoprom](https://github.com/cabify/gotoprom) 5 years ago :D

Reply:
Interesting, will look into this when I have more time.